commit/plugin.go

```diff
@@ -85,8 +101,130 @@ func NewPlugin(
 		reportCodec:       reportCodec,
 		reportingCfg:      reportingCfg,
 		chainSupport:      chainSupport,
-		observer:          observer,
+		merkleProcessor:   merkleProcessor,
+		tokenProcessor:    tokenprice.NewProcessor(),
+		feeProcessor:      chainfee.NewProcessor(),
+	}
+}
+
+func (p *Plugin) Query(_ context.Context, outCtx ocr3types.OutcomeContext) (types.Query, error) {
+	return types.Query{}, nil
+}
+
+func (p *Plugin) ObservationQuorum(_ ocr3types.OutcomeContext, _ types.Query) (ocr3types.Quorum, error) {
+	// Across all chains we require at least 2F+1 observations.
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (p *Plugin) Observation(
+	ctx context.Context, outCtx ocr3types.OutcomeContext, _ types.Query,
+) (types.Observation, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+	fChain := p.ObserveFChain()
+	//TODO: Send fchain instead of computing it inside
+	merkleObs, err := p.merkleProcessor.Observation(ctx, prevOutcome.MerkleRootOutcome, merkleroot.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get merkle observation", "err", err)
+	}
+	tokenPricesObs, err := p.tokenProcessor.Observation(ctx, prevOutcome.TokenPriceOutcome, tokenprice.Query{})
+	if err != nil {
+		//log error
+		p.lggr.Errorw("failed to get token prices", "err", err)
+	}
+	gasObs, err := p.feeProcessor.Observation(ctx, prevOutcome.ChainFeeOutcome, chainfee.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get gas prices", "err", err)
+	}
```

Comment:
Except for the deterministic ordering, this could be a fully decoupled generic SubPlugin aggregator.

Replace:
```
	merkleProcessor   shared.PluginProcessor
	tokenProcessor    shared.PluginProcessor
	feeProcessor      shared.PluginProcessor
```

With:
```
subPlugins map[string]shared.SubPlugin Observation function doesn't change when you add another SubPlugin.
```

This loop becomes:
```
var observations map[string]any
for k, processor := range subPlugins {
    // I suppose args need a little more thought. You could have shared k/v's
    // Not sure about the query.
    observations[k] = processor.Observation(ctx, prevOutcome, query)
}
return Encode(observations)
```

The neat part here is since you're using polymorphism, the

Reply:
I thought about this earlier and didn't go this way to make it simpler, different types will make it more complex once you start having conditions or other mapping for different arguments.

Reply:
We don't need to do it in this PR, but I think its a natural progression of this work.